library/extractor/src/main/java/com/google/android/exoplayer2/text/ssa/SsaDecoder.java

```diff
@@ -376,11 +348,77 @@ private static Cue createCue(
       cue.setPosition(styleOverrides.position.x / screenWidth);
       cue.setLine(styleOverrides.position.y / screenHeight, LINE_TYPE_FRACTION);
     } else {
-      // TODO: Read the MarginL, MarginR and MarginV values from the Style & Dialogue lines.
       cue.setPosition(computeDefaultLineOrPosition(cue.getPositionAnchor()));
       cue.setLine(computeDefaultLineOrPosition(cue.getLineAnchor()), LINE_TYPE_FRACTION);
     }
 
+    // Apply margins if there are no overrides and we have valid positions.
+    if (styleOverrides.alignment == SsaStyle.SSA_ALIGNMENT_UNKNOWN
+        && styleOverrides.position == null
+        && cue.getPosition() != Cue.DIMEN_UNSET
+        && cue.getLine() != Cue.DIMEN_UNSET) {
+      // Margin from Dialogue lines takes precedence over margin from Style line.
+      cue.setPosition(cue.getPosition() + (marginLeft != 0f
```

Comment:
I think adding the left margin onto the cue position only makes sense if the cue is being 'measured' from the left edge.

i.e. `alignment` is one of `[1, 2, 4, 5, 7, 8]` (left or centre)

If the text is right-aligned, then the cue 'start' position is unaffected by a left margin. The only thing that should be affected is the cue width (`Cue.size`), which is currently unset in this class. I think this PR will need to set `Cue.size` in order to respect `MarginL` when alignment is right or centre, and `MarginR` when alignment is left or centre.

In the same way, `MarginV` needs to respect the vertical part of `Alignment`. The spec I'm reading says:

> Field 16:	MarginV. This defines the vertical Left Margin in pixels.
> For a subtitle, it is the distance from the bottom of the screen.
> For a toptitle, it is the distance from the top of the screen.
> For a midtitle, the value is ignored - the text will be vertically centred

Reply:
> If the text is right-aligned, then the cue 'start' position is unaffected by a left margin

Yes, you are right.

> The only thing that should be affected is the cue width (Cue.size)
Can you please elaborate on this? Why should we change the cue width in this case instead of doing nothing? :)

> In the same way, MarginV needs to respect the vertical part of Alignment
Yes, I've missed this.

Reply:
> > The only thing that should be affected is the cue width (Cue.size)
> 
> Can you please elaborate on this? Why should we change the cue width in this case instead of doing nothing? :)

If you consider a subtitle that's left aligned but really long, then without any margin settings you'd expect it to start on the left hand side of the screen, continue to the right, and then (because it's **really** long) wrap onto a second line that starts back on the LHS of the screen again.

If the left margin is set, then updating the position does the right thing here - it will start from an indented position, continue all the way to RHS and wrap back to the same indent.

If the right margin is set, then things start to go wrong without setting `Cue.size`, because (I assume) the person setting the right margin will expect a long subtitle to wrap **at the margin** and not at the edge of the screen. This can be achieved by setting `Cue.size`, and essentially making the cue box shorter than the available screen width - which will cause the long line to wrap 'early' and respect the margin.

The same argument can be made in reverse for right-aligned subtitles. And center-alignment too really.

Reply:
It should be fixed now.